src/System.Management.Automation/engine/regex.cs

```diff
@@ -258,24 +258,33 @@ public static bool ContainsWildcardCharacters(string pattern)
             }
 
             bool result = false;
+            int bracket = 0;
 
             for (int index = 0; index < pattern.Length; ++index)
             {
                 if (IsWildcardChar(pattern[index]))
                 {
                     result = true;
-                    break;
                 }
 
-                // If it is an escape character then advance past
-                // the next character
-
-                if (pattern[index] == escapeChar)
+                switch (pattern[index])
                 {
-                    ++index;
+                    case '[':
+                        ++bracket;
+                        break;
+
+                    case ']':
+                        --bracket;
+                        break;
```

Comment:
@kwkam By looking at the code, as long as the given string contains unpaired `[` or `]`, it's considered not having wildcard characters, even if the string contains `*` or `?`, is that what you are after? If that's the case, you probably shouldn't call `IsWildcardChar`, but use `switch` only instead:
```c#
switch (pattern[index])
{
    case '*':
    case '?':
        result = true; break;
    case '[':
        result = true;
        ++bracket;
        break;
    case ']':
        result = true;
        --bracket;
        break;
    case escapeChar:
        ++index;
        break;
```
But still, this is not good enough, what if you have even number of `[` and `]` but the order is wrong? Like `./][.ps1`?

Reply:
Yeah, this is a context-free grammar -- you'll need a stack to recognise it.

Reply:
Reverted.